### ISSUE TITLE
Iteration over ranges with non-param strides result in runtime check for ambiguity 

### ISSUE THREAD

**Issue by e-kayrakli** [chapel-lang/chapel#22176]
```chapel
var myStride = 3;

for i in 1..10 by myStride {
  writeln("hello");
}
```

will have runtime checks coming from places like [`alignedLowAsInt`](https://github.com/chapel-lang/chapel/blob/main/modules/internal/ChapelRange.chpl#L683-L685) while I can see that we need to check that, I am not understanding (a) why this is not ambiguous, (b) why can't we check this at compile time. https://github.com/chapel-lang/chapel/issues/19717 could be a proposal that address this, but I can't be sure.

Similarly, I was hoping that I can work around that check by explicitly aligning my range:

```chapel
var myStride = 3;

for i in 1..10 by myStride align 1{
  writeln("hello");
}
```

this doesn't change the behavior.

Naively, I tried to add a new param overload of [`isAmbiguous`](https://github.com/chapel-lang/chapel/blob/main/modules/internal/ChapelRange.chpl#L824-L831) that has `where stridable && aligned`, but unfortunately whether a range has `align` is not a compile-time property. This may be because we don't want it to be part of the type of the range, but currently there's no way of having a `param` property that doesn't change the type signature.

---

The problem at my hand that pushed me towards this is from GPUs. If you have a `for` over a strided range like the ones above inside a `foreach`, it'll thwart GPU offload because of the runtime check potentially leading to `halt`. You can't `halt` from GPUs today. Arguably, this particular problem's solution is to get `halt`s to work on GPUs. But this particular case is the first time we're seeing the lack hurting usability.

**Comment by e-kayrakli** [comment]
@vasslitvinov -- something like this might have come up during range review, but I can't find any issue on a quick look.
@bradcray -- I think you have also done quite a lot of work in this area, so I'm wondering what your take is.

**Comment by vasslitvinov** [comment]
@e-kayrakli you are correct in your evaluation. The outcome of the discussion of the proposals to make `aligned` a compile-time property, at least in some cases was that we did not see the benefit of the added complexity [needed to make aligned a compile-time property in some cases] or of making things work differently than they do today [needed to make aligned a compile-time property in all cases]. Reference: https://github.com/chapel-lang/chapel/issues/19717#issuecomment-1277037842

**Comment by vasslitvinov** [comment]
Thinking about it: if we cannot enable halts on GPUs and are still unwilling to make changes to `aligned` semantics: another option is to add a compiler analysis that removes isAmbiguous checks for results of a `by` operation on bounded unstridable ranges.

Aside: the isAmbiguous check is/should probably be disabled upon --no-checks. Yet, of course we want GPUs work without --no-checks.

**Comment by e-kayrakli** [comment]
> Aside: the isAmbiguous check is/should probably be disabled upon --no-checks. Yet, of course we want GPUs to work without --no-checks.

Can it? 

https://github.com/chapel-lang/chapel/blob/main/modules/internal/ChapelRange.chpl#L683 in this case, can you get an alignedLow if the alignment is ambiguous? In my understanding, the answer is "no", but my understanding is too wonky. So I am not trying to convince you but more really asking. :)

**Comment by vasslitvinov** [comment]
Correct, requesting `alignedLow` is not a valid operation when the range is ambiguous aka unaligned. My proposal is to follow the lead of array indexing where the bounds checking is disabled upon --no-checks even though indexing out of bounds is invalid.

**Comment by vasslitvinov** [comment]
Looking at it with fresh eyes...

I would like to hide the alignment check in `range.alignedLowAsInt()` and `range.alignedHighAsInt()` behind an `if boundsChecking`. This is a no-brainer to me because of the nature of the check. This is not sufficient to enable GPU offloading for a "foreach" over a strided range, however it is a step forward.

I would like to prove that a bounded range cannot be ambiguous, er, unaligned. The case to consider is slicing where one or both ranges are unbounded and/or unaligned. (Other challengers, anyone?) Once this is established, we extend the param overload `range.hasParamAligned()` to include bounded ranges, and boom, alignedLowAsInt / alignedHighAsInt will no longer call halt() in the cases at hand.

Iteration over ranges with param strides IS offloaded go GPUs however the stride is ignored. It may be the case that the stride would be ignored for non-param strides, once those get offloaded. This is a bug for us to fix.

**Comment by e-kayrakli** [comment]
> This is not sufficient to enable GPU offloading for a "foreach" over a strided range, however it is a step forward.
> 
> Iteration over ranges with param strides IS offloaded go GPUs however the stride is ignored. It may be the case that the stride would be ignored for non-param strides, once those get offloaded. This is a bug for us to fix.

I think these are connected, right? Our GPU transformer may be ignoring stride and I agree that we should fix that. However, where this was more worrisome for me was:

```chpl
on here.gpus[0] {
  foreach i in 1..10 { // benign loop header in terms of GPU eligibility
    for j in 1..20 by myStride {  // whoops, can't GPUize the outer loop
                                  // because _the body_ of that loop now has
                                  // halt because of this loop
```

This pattern is the most intuitive pattern to to make GPU threads handle more than one iteration without https://github.com/chapel-lang/chapel/issues/22175 (which I know you're working on)
